fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/villager/VillagerTypeFactory.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.object.builder.v1.villager;
+
+import net.minecraft.item.Item;
+import net.minecraft.util.Identifier;
+import net.minecraft.village.VillagerType;
+
+import net.fabricmc.fabric.impl.object.builder.FabricVillagerType;
+
+public final class VillagerTypeFactory {
+	private VillagerTypeFactory() {
+	}
+
+	/**
+	 * Creates a Villager type.
+	 *
+	 * <p>Note that all villager types should be created and registered before any trade offers are created.
+	 * If trade offers are added before all villager types have been registered, the game may crash due to how minecraft handles {@link net.minecraft.village.TradeOffers.TypeAwareBuyForOneEmeraldFactory type aware trades}.
+	 *
+	 * @param id the id of the villager type.
+	 * @param fishermanBoatItem the boat item this villager type prefers if the villager's profession is a {@link net.minecraft.village.VillagerProfession#FISHERMAN fisherman}.
+	 *
+	 * @return a new villager type.
+	 */
+	public static VillagerType create(Identifier id, Item fishermanBoatItem) {
```

Comment:
![concern](https://cdn.discordapp.com/emojis/707016186281590855.png?v=1)
You cannot make sure that mods don't add more trades that has type specific items.

Reply:
So how do we work around this issue then of mods being able to use the type aware factory? The item being passed is for that fisherman `1 boat -> 1 emerald` trade, which mojang modifies the boat being bought based on the villager type.